processor/src/test/java/org/mapstruct/ap/test/targetthis/CustomerEntityMapper1.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.targetthis;
+
+import org.mapstruct.CollectionMappingStrategy;
+import org.mapstruct.InheritInverseConfiguration;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+import org.mapstruct.ReportingPolicy;
+import org.mapstruct.factory.Mappers;
+
+@Mapper(
+        unmappedSourcePolicy = ReportingPolicy.IGNORE,
+        unmappedTargetPolicy = ReportingPolicy.IGNORE,
+        collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED
+)
+public interface CustomerEntityMapper1 {
+    CustomerEntityMapper1 INSTANCE = Mappers.getMapper( CustomerEntityMapper1.class );
+
+    @Mapping(target = ".", source = "entity")
+    CustomerEntity map(CustomerDTO customer);
+
+    @InheritInverseConfiguration(name = "map")
+    // @Mapping(target = "entity", source = ".")
+    CustomerDTO inverseMap(CustomerEntity customer);
+
+    @Mapping(target = "entity", source = ".")
```

Comment:
its currently (without this solution) possible to specify as source: `parameter`, `parameter.property` and `property` (omitting `parameter` altogether).  If you have multiple sources I think ~~(need to check)~~ the parameter designation is mandatory: so `parameter1.propX`, `parameter2.propY`. BTW: this is stuff inside `SourceReference`. How does `"."` relate to all the other options on source side? Will it help or confuse the user to have an additional `source=` option?

Reply:
you probably need an additional test with multiple source arguments. So something like:
```java 
@Mapping( target = ".", source = "param1" ) // first set of properties, lets say "propX", "propY"
@Mapping( target = ".", source= "param2" ) // second set of properties, lets say "propU", "propW"
Target map( Source1 param1, Source2 param2 );
```

and

```java 
@Mapping( target = ".", source = "param1" ) // first set of properties, lets say "propX", "propY"
@Mapping( target = ".", source= "param2" ) // second set of properties, lets say "propU", "propW" 
void map( @MappingTarget Target target, Source1 param1, Source2 param2 );
```

Reply:
Thinking of that a bit more.. Perhaps we should allow `"."`  ony in one specific case:  when we have no `@MappingTarget` to indicate the target (then the user could refer to the `@MappingTarget` annotated parameter.  This to avoid ambiguity.

We should not allow it in the source mapping attribute **at all** . So we probably have to do some checking (+ unit test) when parsing the mirror in the `[Mapping](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/model/source/Mapping.java)` class. Also, we should take care of the `@InheritReverseConfiguration` (`Mapping#invert")? to strip this..

So this is ok:

```java 
@Mapping( target = "." ) 
Target map( Source1 param );
```
But, this is **not ok**:
```java 
@Mapping( target = "." ) 
void map( Source1 param, @MappingTarget Target target );
```
```java 
@Mapping( target = "." ) 
void Target map( Source1 param, @MappingTarget Target target );
```
```java 
@Mapping( target = "target.property", source = "." ) 
Target map( Source1 param );
```

Reply:
I think most common case for us is
`
@Mapping( target = ".", source = "param1" )
@Mapping( target = ".", source= "param2" )
Target map( SourceObject source  );
`

but I can see that sometimes I would like to use something like 
`
@Mapping( target = ".", source = "param1" )
@Mapping( target = ".", source= "param2.innerproperty" )
Target map( SourceObject source );
`
and it's probably difficult to have "innerproperty" as parameter (just guessing).

Also what will happen if you have some property as target and target mapping?
`
@Mapping( target = "name" ) 
void map( Source1 param, @MappingTarget Target target );
`

In my mind information in @Mapping is guidance from direct mapping Source.property -> Target.property. And in such way it's Ok to have @Mapping and MappingTarget, as:

`
@Mapping( target = ".", source = "param1" )
@Mapping( target = ".", source= "param2.innerproperty" )
public void map( SourceObject source, @MappingTarget Target target );
`
To be honest I haven't tested that, but for me it doesn't create any ambiguity because I see it as "set all properties from Source.param1 to current target (target), set all properties from innerproperty that is nested property on param2 that is property of Source to current target (target) and map all other properties from source to target)". So in this case only difference I can see is that Target is moved from return parameter to update parameter (and that of course will force me to create object myself first).
When there are more than one source parameters it gets a bit ambiguous  in such way, that I might not be sure if param1 belongs to source1 and param to belongs to source2, or they can be in any of these objects, but I would have some thing with named target parameter anyway.

And just to be on same page, target="." without source makes no sense for me, because that's default behaviour anyway (mapping all properties from source to target).

Reply:
btw I wonder about :

```
void map( @MappingTarget Target target, Source1 param1, Source2 param2 );
```

what happens when param1 and param2 have property with same name, or they need to be mapped to different target properties (param1.name -> target.name, param2.name -> target.title)?

Reply:
@dainiusfigoras .. Its a lot of text to gauge.. But lets try to setup a small convention.. 

* With `param` I mean a parameter (argument) of the mapping method. 

So with respect to a mapping method you have:
1. Always one target parameter (`@MappingTarget`) and / or return
2. One or more source parameters
3. One or more (@Context) annotated context parameters

* With `property` I mean a property in as in `bean properties`, exposed via a readAccessor / writeAccessor (normally getter / setter)

We try to make use of that consistently, when working on the source code and on examples. Take for instance a look  `SourceReference` / `TargetReference` in the source code of MapStruct.

---

Now, I am concerned about complexity.. The API should be clear. That should always have our focus. It can't be one use case that we try to implement opens up a can of worms for all kinds of other use cases (believe me, this happens :smile: ) that we need to correct with patches.

 I'll try to explain by my examples what I meant case by case:

> But, this is **not ok**:

**Case 1:**

> 
> ```java
> @Mapping( target = "." ) 
> void map( Source1 param, @MappingTarget Target target );
> ```

Here the user already has a possibility to refer to the target  by means of `target="target"`. Opening up more possibilities increases complexity. 

- if we detect the setup above we should give an error (fail fast)
- if we encounter an `@InheritConfiguration` from a mapping method with a "return type only" like this:

```java
@InheritConfiguration( name = "map" ) // the method above
Target map( Source1 param );
```
we should handle this correctly in the source code (`Mapping#targetReference`) I guess. 

 **Case 2:**
> ```java
> @Mapping( target = "." ) 
> void Target map( Source1 param, @MappingTarget Target target );
> ```
> 
- if we detect the setup above we should give an error (fail fast)
- same reasoning as above. We have a way to indicate the target by a FQN (`target="target"`) because the `@MappingTarget` identical to the return type

**Case 3:**
> ```java
> @Mapping( target = "target.property", source = "." ) 
> Target map( Source1 param );
> ```
- if we detect the setup above we should give an error (fail fast)
- Allowing `source="."` opens a new possibility next the ones we already have (see my explanation above). So let's not do that.  In case of a method annotated with `@InheritInverseConfiguration` we should adapt in our source code (`Mapping#inverse`) to point to the write source parameter. So example:
```java
@Mapping( target = ".", source = "param" ) 
Target map( Source1 param );

@InheritInverseConfiguration // from the method above
Source map( Target param);
```

Reply:
> I think most common case for us is
> 
> ```java
> @Mapping( target = ".", source = "param1" )
> @Mapping( target = ".", source= "param2" )
> Target map( SourceObject source  );
> ```

I guess you mean 
```java
 @Mapping( target = ".", source = "property1" )
@Mapping( target = ".", source= "property2" )
Target map( SourceObject source  );
```
Or identical to this (fully qualifying parameter as well): 

```java
 @Mapping( target = ".", source = "source.property1" )
@Mapping( target = ".", source= "source.property2" )
Target map( SourceObject source  );
```

> but I can see that sometimes I would like to use something like
> 
> ```java
> @Mapping( target = ".", source = "param1" )
> @Mapping( target = ".", source= "param2.innerproperty" )
> Target map( SourceObject source );
> ```
```java
 @Mapping( target = ".", source = "property1" )
@Mapping( target = ".", source= "property2.innerProperty" )
Target map( SourceObject source  );
```
?
Yes that should be supported as well, meaning that all properties (except `property1.innerproperty`  should be mapped from level `property1` combined with all properties from `property1.innerproperty` 

> and it's probably difficult to have "innerproperty" as parameter (just guessing).
> 
Not sure what you mean here. 

> Also what will happen if you have some property as target and target mapping?
> 
> ```java
> @Mapping( target = "name" ) 
> void map( Source1 param, @MappingTarget Target target );
> ```
Not sure what this would mean / or what the user would intend with this.. I guess MapStruct would give an error.  (Nice to try BTW). But cases with only a parameter are possible.. For instance:
 ```java
@Mapping( target = "name", qualfiedBy = "SomeQualfier.class"  ) 
void map( Source1 param, @MappingTarget Target target );
```
Meaning that there would be a `param.name` as present as well in the source that can be mapped based on its name. 
Meaning: the

Reply:
> In my mind information in @mapping is guidance from direct mapping Source.property -> Target.property.

`@Mapping` is a guidance where MapStruct cannot make a decision on its own. For instance, when property names differ, or when there's a conflict in which method / conversion to select.

Reply:
> ```java
> @Mapping( target = ".", source = "param1" )
> @Mapping( target = ".", source= "param2.innerproperty" )
> public void map( SourceObject source, @MappingTarget Target target );
> ```
> 
> To be honest I haven't tested that, but for me it doesn't create any ambiguity because I see it as "set all properties from Source.param1 to current target (target), set all properties from innerproperty that is nested property on param2 that is property of Source to current target (target) and map all other properties from source to target)". So in this case only difference I can see is that Target is moved from return parameter to update parameter (and that of course will force me to create object myself first).
> When there are more than one source parameters it gets a bit ambiguous in such way, that I might not be sure if param1 belongs to source1 and param to belongs to source2, or they can be in any of these objects, but I would have some thing with named target parameter anyway.

I hope I made my point clear above. It is not about the workings of such mappings.. The simple reason I don't like to have this is because you can also specify:
```java
@Mapping( target = "target", source = "param1" )
@Mapping( target = "target", source= "param2.innerproperty" )
public void map( SourceObject source, @MappingTarget Target target );
 ```
Or alternatively:
```java
@Mapping( target = "target", source = "source.param1" )
@Mapping( target = "target", source= "source.param2.innerproperty" )
public void map( SourceObject source, @MappingTarget Target target );
 ```

Reply:
> Case3 should not work and give compile error.
I would say: a MapStruct error.. It should be detected in: `Mapping#fromMappingPrism` (there are already many of such checks there). 

> It sounds a bit dangerous because that syntax is working in current release and removing it would mean changing existing behaviour.
How do you mean? I was not aware I could currently specify `source="."`?

Reply:
> How do you mean? I was not aware I could currently specify `source="."`?

I checked. and you're right.. Seems I don't know all capabilities of MapStruct :smile: 

`org.mapstruct.ap.test.nestedsource.parameter.LetterMapper` changed: 

```java
    @InheritInverseConfiguration
    @Mapping(target = "font", source = "entity")
    LetterDto deNormalizeLetter(LetterEntity entity);
```
into
```java
    @InheritInverseConfiguration
    @Mapping(target = "font", source = ".")
    LetterDto deNormalizeLetter(LetterEntity entity);
```
and it gives exactly the same code..

Reply:
> Seems I don't know all capabilities of MapStruct 😄

it's more powerful then you can imagine :D ;)